File: dsa/concepts/01_arrays/java/L414_ThirdMaximumNumber.java
public class L414_ThirdMaximumNumber {

	public static void main(String[] args) {
		L414_ThirdMaximumNumber obj = new L414_ThirdMaximumNumber();

		int[] nums = new int[] {-2147483648,-2147483648,-2147483648,-2147483648,1,1,1};
		System.out.println(obj.thirdMax(nums));
	}

	/*
	 * Problem: https://leetcode.com/problems/third-maximum-number/
     * Idea: Swap max -> secMax, secMax -> thirdMax
	 * */
	public int thirdMax(int[] nums) {
        if(nums.length == 1) return nums[0];
        if(nums.length == 2) return Math.max(nums[0], nums[1]);

        Integer max = null, secondMax = null, thirdMax = null;
        for(Integer num: nums){
            if(num.equals(max) || num.equals(secondMax) || num.equals(thirdMax)) continue;

            if(max == null || num > max){
                thirdMax = secondMax;   //n
                secondMax = max;        //-1
                max = num;              //1
            }else if(secondMax == null || num > secondMax){
                thirdMax = secondMax;
                secondMax = num;
            }else if(thirdMax == null || num > thirdMax){
                thirdMax = num;
            }
        }

        if(thirdMax != null ) return thirdMax;
        return max;
    }
}

File: dsa/patterns/01_arrays/python/l414_third_maximum_number.py
"""
L414 - Third Maximum Number

Problem: https://leetcode.com/problems/third-maximum-number/
Idea: Swap max -> secMax, secMax -> thirdMax
"""

def third_max(nums):
    if len(nums) == 1:
        return nums[0]
    if len(nums) == 2:
        return max(nums[0], nums[1])
    
    max_val = None
    second_max = None
    third_max = None
    
    for num in nums:
        if num == max_val or num == second_max or num == third_max:
            continue
        
        if max_val is None or num > max_val:
            third_max = second_max  # n
            second_max = max_val    # -1
            max_val = num           # 1
        elif second_max is None or num > second_max:
            third_max = second_max
            second_max = num
        elif third_max is None or num > third_max:
            third_max = num
    
    if third_max is not None:
        return third_max
    return max_val


if __name__ == "__main__":
    nums = [-2147483648, -2147483648, -2147483648, -2147483648, 1, 1, 1]
    print(third_max(nums))
